nat.ts

```diff
@@ -28,8 +28,8 @@ module J2ME {
   import Bytecodes = Bytecode.Bytecodes;
   import toHEX = IntegerUtilities.toHEX;
 
-  export function asyncImplOld(returnKind: string, promise: Promise<any>) {
-    return asyncImpl(kindCharacterToKind(returnKind), promise);
+  export function asyncImplOld(returnKind: string, promise: Promise<any>, cleanup?: Function) {
```

Comment:
Another reason why I added the cleanup function is that it needs to be called after the reference is in the asm.js heap (so after the line *i32[sp++] = l;*).
In practice, we need to avoid running any Java code between the call to _gcFree and the line I mentioned above (otherwise a GC could happen in the meantime freeing the array).